### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by andreypopp**
Sorry for such a huge PR—I was hoping to get things into "almost working" state quick and then iterate with more granular changesets.

What's in there:
* **refactoring:** `esy build` renamed to `esy build-eject` as it only prints a generated `Makefile`
* **refactoring:** `esy build` is now effectively `make -f <(esy build-eject) build`
* **new command:** `esy shell` (`make -f <(esy build-eject) shell`) spawns a new shell with env set for the sandbox. Note that ejected `Makefile` has also `<pkg name>.shell` targets for each package in a sandbox. Didn't expose them through `esy` command yet though.
* **new command:** `esy clean` (`make -f <(esy build-eject) clean`) removes `_install` and `_build` trees.
* **new command:** `esy <args>` now executes `<args>` in a sandbox environment. For example: use `esy ocaml` to run REPL in a sandbox.
* **refactoring:** `lib/PackageDb.js` represents sandbox state (packages and deps between them)
* **refactoring:** Move `.bin/esyEnv.js` to `lib/PackageEnvironment.js`

### CODE REVIEW

**Comment by andreypopp** [comment]
Forgot to mention: [esy-ocaml-project](https://github.com/andreypopp/esy-ocaml-project) now serves as a minimal example of using `esy`.

File: lib/esyBuildEjectCommand.js [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+            `${packageJson.name}.findlib.conf`,
+            ...dependencies
+          ],
+          command: isRootPackage
+          ? outdent`
+            $(ESY__PREPARE_CURRENT_INSTALL_TREE)\\
+            ${buildCommand}
+          `
+          : outdent`
+            if [ ! -d "$cur__install" ]; then \\
```
Comment (andreypopp) [inline lib/esyBuildEjectCommand.js:201]: So this is super silly, I know. We need to proper build caching instead (using env and deps' env as key probably). But for now it is at least fast. Note that it still always rebuilds root package. Also there's `<pkg name>.rebuild` target in generated `Makefile` to force rebuild of the package.

File: lib/esyBuildEjectCommand.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+          destdir = "${installDir(packageJson.name, 'lib')}"
+        `
+      });
+
+      // Produce macro with rendered package's environment.
+      prelude.push({
+        type: 'define',
+        name: `${normalizedName}__ENV`,
+        value: Makefile.renderEnv(buildEnvironment),
+      });
```
Comment (andreypopp) [inline lib/esyBuildEjectCommand.js:152]: Right now we dump entire env for a package in a single `define`. Instead we could reduce the size of the `Makefile` by dumping each group into separate `define` and then combine them together for each package. I think this is important to keep the generated `Makefile` readable.

File: lib/esyBuildEjectCommand.js [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        target: 'findlib.conf',
+        dependencies: [buildDir(packageJson.name, 'findlib.conf')],
+      }));
+
+      rules.push(makePackageRule({
+        target: 'shell',
+        command: outdent`
+          $SHELL \\
+            --noprofile \\
+            --rcfile <(echo 'export PS1="[$cur__name sandbox] $ "')
```
Comment (andreypopp) [inline lib/esyBuildEjectCommand.js:175]: `PS1` needs to be more fancy.

File: package.json [hunk 4]
```diff
@@ -7,12 +7,12 @@
     "test": "echo \"Error: no test specified\" && exit 1"
   },
   "bin": {
-    "esy": ".bin/esy",
-    "esy.js": ".bin/esy.js",
-    "esyEnv.js": ".bin/esyEnv.js"
+    "esy": ".bin/esy"
   },
   "dependencies": {
+    "flow-remove-types": "^1.0.5",
```
Comment (andreypopp) [inline package.json:13]: This is runtime dep for now, need to precompile sources to strip flow before publishing. At least for now it is easier to experiment with `esy` that way.

The best would be to rewrite `esy` in OCaml/Reason.

File: lib/PackageDb.js [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+  handler(packageJsonPathOnEjectingHost, pkg);
+}
+
+function normalizeName(name) {
+  return name
+    .toLowerCase()
+    .replace(/@/g, '')
+    .replace(/\//g, '_')
+    .replace(/\-/g, '_');
+}
```
Comment (andreypopp) [inline lib/PackageDb.js:254]: So normalised package name is lower cased as well as built-ins var names, examples:
* `$cur__root`
* `$packagea__lib`
* `$packagea__custom_var`
* ...

**Comment by andreypopp** [comment]
Let me know if you want this PR to be split into several less huge ones.

**Comment by jordwalke** [comment]
Thank you. I'm excited about this.

**Pull request merged**